auto-configurations/mcp/spring-ai-autoconfigure-mcp-client-common/src/main/java/org/springframework/ai/mcp/client/common/autoconfigure/annotations/McpSyncAnnotationCustomizer.java

```diff
@@ -0,0 +1,179 @@
+	}
+
+	@Override
+	public void customize(String name, SyncSpec clientSpec) {
+
+		if (!CollectionUtils.isEmpty(this.syncElicitationSpecifications)) {
+			this.syncElicitationSpecifications.forEach(elicitationSpec -> {
+				Stream.of(elicitationSpec.clients()).forEach(clientId -> {
+					if (clientId.equalsIgnoreCase(name)) {
+						// Check if client already has an elicitation spec
+						if (this.clientElicitationSpecs.containsKey(name)) {
+							throw new IllegalArgumentException("Client '" + name
+									+ "' already has an elicitationSpec registered. Only one elicitationSpec is allowed per client.");
+						}
+
+						this.clientElicitationSpecs.put(name, Boolean.TRUE);
+						clientSpec.elicitation(elicitationSpec.elicitationHandler());
+
+						logger.info("Registered elicitationSpec for client '{}'.", name);
+					}
+				});
+			});
+		}
+
+		if (!CollectionUtils.isEmpty(this.syncSamplingSpecifications)) {
+			this.syncSamplingSpecifications.forEach(samplingSpec -> {
+				Stream.of(samplingSpec.clients()).forEach(clientId -> {
+					if (clientId.equalsIgnoreCase(name)) {
+
+						// Check if client already has a sampling spec
+						if (this.clientSamplingSpecs.containsKey(name)) {
+							throw new IllegalArgumentException("Client '" + name
+									+ "' already has a samplingSpec registered. Only one samplingSpec is allowed per client.");
+						}
+						this.clientSamplingSpecs.put(name, Boolean.TRUE);
+
+						clientSpec.sampling(samplingSpec.samplingHandler());
+
+						logger.info("Registered samplingSpec for client '{}'.", name);
+					}
+				});
+			});
+		}
+
+		if (!CollectionUtils.isEmpty(this.syncLoggingSpecifications)) {
+			this.syncLoggingSpecifications.forEach(loggingSpec -> {
+				Stream.of(loggingSpec.clients()).forEach(clientId -> {
+					if (clientId.equalsIgnoreCase(name)) {
+						clientSpec.loggingConsumer(loggingSpec.loggingHandler());
+						logger.info("Registered loggingSpec for client '{}'.", name);
+					}
+				});
+			});
+		}
+
+		if (!CollectionUtils.isEmpty(this.syncProgressSpecifications)) {
+			this.syncProgressSpecifications.forEach(progressSpec -> {
+				Stream.of(progressSpec.clients()).forEach(clientId -> {
+					if (clientId.equalsIgnoreCase(name)) {
+						clientSpec.progressConsumer(progressSpec.progressHandler());
+						logger.info("Registered progressSpec for client '{}'.", name);
+					}
+				});
+			});
+		}
+
+		if (!CollectionUtils.isEmpty(this.syncToolListChangedSpecifications)) {
+			this.syncToolListChangedSpecifications.forEach(toolListChangedSpec -> {
+				Stream.of(toolListChangedSpec.clients()).forEach(clientId -> {
+					if (clientId.equalsIgnoreCase(name)) {
+						clientSpec.toolsChangeConsumer(toolListChangedSpec.toolListChangeHandler());
+						logger.info("Registered toolListChangedSpec for client '{}'.", name);
+					}
+				});
+			});
+
+			if (!CollectionUtils.isEmpty(this.syncResourceListChangedSpecifications)) {
```

Comment:
is the 'if' clause (and the one below) indented to be part of line 145 if statement?    i think not, but i'll just highlight it here due to the async nature of the where we are and fix later as this is a big PR and I don't want to make a change without checking on this first.

Reply:
Ah, the async version doesn't have this, so i'll fix.